### ISSUE TITLE
App Engine Maven Plugin fails w/ Process failed with exit code: 1

### ISSUE THREAD

**Issue by saturnism** [GoogleCloudPlatform/app-maven-plugin#409]
Might be similar issue w/ Cloud Code issue: https://github.com/GoogleCloudPlatform/cloud-code-intellij/issues/2662

```
<plugin>
  <groupId>com.google.cloud.tools</groupId>
  <artifactId>appengine-maven-plugin</artifactId>
  <version>2.2.0</version>
</plugin>
```

```
[ERROR] Failed to execute goal com.google.cloud.tools:appengine-maven-plugin:2.2.0:deploy (default-cli) on project demo-appengine: Execution default-cli of goal com.google.cloud.tools:appengine-maven-plugin:2.2.0:deploy failed: com.google.cloud.tools.managedcloudsdk.ManagedSdkVerificationException: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:appengine-maven-plugin:2.2.0:deploy (default-cli) on project demo-appengine: Execution default-cli of goal com.google.cloud.tools:appengine-maven-plugin:2.2.0:deploy failed: com.google.cloud.tools.managedcloudsdk.ManagedSdkVerificationException: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal com.google.cloud.tools:appengine-maven-plugin:2.2.0:deploy failed: com.google.cloud.tools.managedcloudsdk.ManagedSdkVerificationException: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
Caused by: java.lang.RuntimeException: com.google.cloud.tools.managedcloudsdk.ManagedSdkVerificationException: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkDownloader.downloadIfNecessary (CloudSdkDownloader.java:88)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.buildCloudSdk (CloudSdkAppEngineFactory.java:142)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.buildCloudSdkMinimal (CloudSdkAppEngineFactory.java:100)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.getGcloud (CloudSdkAppEngineFactory.java:159)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.newConfigReader (CloudSdkAppEngineFactory.java:208)
    at com.google.cloud.tools.maven.deploy.Deployer$Factory.newDeployer (Deployer.java:41)
    at com.google.cloud.tools.maven.deploy.DeployMojo.execute (DeployMojo.java:35)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
Caused by: com.google.cloud.tools.managedcloudsdk.ManagedSdkVerificationException: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1
    at com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk.isUpToDate (ManagedCloudSdk.java:163)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkDownloader.downloadIfNecessary (CloudSdkDownloader.java:76)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.buildCloudSdk (CloudSdkAppEngineFactory.java:142)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.buildCloudSdkMinimal (CloudSdkAppEngineFactory.java:100)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.getGcloud (CloudSdkAppEngineFactory.java:159)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.newConfigReader (CloudSdkAppEngineFactory.java:208)
    at com.google.cloud.tools.maven.deploy.Deployer$Factory.newDeployer (Deployer.java:41)
    at com.google.cloud.tools.maven.deploy.DeployMojo.execute (DeployMojo.java:35)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
Caused by: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1
    at com.google.cloud.tools.managedcloudsdk.command.CommandCaller.call (CommandCaller.java:68)
    at com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk.isUpToDate (ManagedCloudSdk.java:152)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkDownloader.downloadIfNecessary (CloudSdkDownloader.java:76)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.buildCloudSdk (CloudSdkAppEngineFactory.java:142)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.buildCloudSdkMinimal (CloudSdkAppEngineFactory.java:100)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.getGcloud (CloudSdkAppEngineFactory.java:159)
    at com.google.cloud.tools.maven.cloudsdk.CloudSdkAppEngineFactory.newConfigReader (CloudSdkAppEngineFactory.java:208)
    at com.google.cloud.tools.maven.deploy.Deployer$Factory.newDeployer (Deployer.java:41)
    at com.google.cloud.tools.maven.deploy.DeployMojo.execute (DeployMojo.java:35)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

**Comment by saturnism** [comment]
Similar stackoverflow issue:
https://stackoverflow.com/questions/52535556/error-while-executing-appengines-downloadcloudsdk-task-in-my-subproject

**Comment by saturnism** [comment]
After debugging app engine core, I was able to get more information on what caused the failure:

```
ERROR: gcloud failed to load: No module named cache
    gcloud_main = _import_gcloud_main()
    import googlecloudsdk.gcloud_main
    from googlecloudsdk.calliope import base
    from googlecloudsdk.calliope import display
    from googlecloudsdk.core.cache import cache_update_ops

This usually indicates corruption in your gcloud installation or problems with your Python interpreter.

Please verify that the following is the path to a working Python 2.7 executable:
    /Users/rayt/homebrew/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python

If it is not, please set the CLOUDSDK_PYTHON environment variable to point to a working Python 2.7 executable.

If you are still experiencing problems, please reinstall the Cloud SDK using the instructions here:
    https://cloud.google.com/sdk/

	at com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk.isUpToDate(ManagedCloudSdk.java:163)
	at com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk.main(ManagedCloudSdk.java:250)
Caused by: com.google.cloud.tools.managedcloudsdk.command.CommandExitException: Process failed with exit code: 1
	at com.google.cloud.tools.managedcloudsdk.command.CommandCaller.call(CommandCaller.java:68)
	at com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk.isUpToDate(ManagedCloudSdk.java:152)
	... 1 more

```

**Comment by saturnism** [comment]
`ManagedCloudSdk.getSdkHome()` points to `/.../Library/Application Support/google-cloud-tools-java/managed-cloud-sdk/LATEST/google-cloud-sdk/bin` Which was in fact installed by Cloud Code and/or the plugin.

In GoogleCloudPlatform/cloud-code-intellij#2662 deleting this directory allows successful reinstall of the SDK.

Manually running `gcloud` in this directly failed with the same error message.

Unsure why the installation is corrupted.

**Comment by saturnism** [comment]
Deleteting the `LATEST` directory allowed managed SDK to be reinstalled and now it works again. However it doesn't explain why  the directory became corrupted in the first place. Suspecting that after a version upgrade, the SDK won't be able to run anymore.

**Comment by saturnism** [comment]
Corollary to this, since I have installed `gcloud` via Homebrew, and `gcloud` is already usable, wouldn't it make sense to use this by default?

**Comment by chanseokoh** [comment]
FYI, you can use the `<cloudSdkPath>` parameter to use your `gcloud` installed yourself.

I think we decided to auto-install a manged SDK to reduce as much friction as possible and make it always up-to-date.

**Comment by chanseokoh** [comment]
https://cloud.google.com/appengine/docs/flexible/java/maven-reference

**Comment by briandealwis** [comment]
And `gcloud` installed from a package system can't be auto-updated.

@saturnism was your Python updated?  
```
Please verify that the following is the path to a working Python 2.7 executable:
    /Users/rayt/homebrew/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python
```

Does `gcloud` record which python executable was used?

**Comment by saturnism** [comment]
@chanseokoh I know the `cloudSdkPath` param, but that's not portable at all.  It essentially ties my build to my environment. The choices should be - use locally available gcloud that's in path (default?), or, if not in path, use gcloud from a specific path, or, use a managed gcloud.

If my manually installed gcloud is out dated, it can attempt to update, w/ the same command line. Or, fail w/ a message to update, or change to a managed version.

**Comment by saturnism** [comment]
The `/Users/rayt/homebrew/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python` path is valid and it's `Python 2.7.16`

**Comment by chanseokoh** [comment]
> If my manually installed gcloud is out dated, it can attempt to update, w/ the same command line. Or, fail w/ a message to update, or change to a managed version.

I understand the reasoning. I already have gcloud locally, then can't the plugin just use it instead of downloading a new one?

But gcloud installed by package managers cannot be updated with `gcloud components udpate`. It has to be updated with the package manager. So, the only other option is to fail with a message to manually update whenever gcloud has a new release, which can be a viable strategy, but I am not so sure if leads to a better experience.

And we have to think about whether it makes sense for the plugin to modify the user's gcloud installation silently. The user may have a reason to stick to a certain version and keep it that way.

That said, now I realized the `cloudSdkPath` reference says

```
Optional parameter to configure the location of the Cloud SDK. Default is to automatically discover from the PATH or standard install location.
```

so, I think it should have used gcloud in your `PATH`.

**Comment by saturnism** [comment]
> But gcloud installed by package managers cannot be updated with gcloud components udpate

I use Homebrew and that worked. I guess it would fail w/ other package managers. May need `sudo` :)

> so, I think it should have used gcloud in your PATH.

I thought so too! But it fails when trying to update the managed version. To the question is - which one does it actually use?? If it uses my `gcloud` in `PATH`, then why bother w/ the managed Cloud SDK in the first place?

**Comment by chanseokoh** [comment]
@loosebazooka could this be a bug that the plugin doesn't use the `gcloud` in `PATH`? Or the doc is wrong? (And I think we can rule out the possibility that a managed Cloud SDK might take precedence over a `gcloud` in `PATH`, since @saturnism deleted the managed Cloud SDK and the plugin downloaded it again.)

> I use Homebrew and that worked.

That's interesting. Perhaps `Homebrew` is special. The [doc](https://cloud.google.com/sdk/docs/components) says

> Note: These instructions will not work if you have installed Cloud SDK using a package manager such as APT or yum because Cloud SDK Component Manager is disabled when using that method of installation. Use these [packages](https://cloud.google.com/sdk/docs/components#external_package_managers) instead.

which matches my past experience. But even requiring `sudo` is a deal breaker.

**Comment by loosebazooka** [comment]
Ever since we started using the managed cloud sdk I think we removed (across all tools, gradle/maven/intellij/eclipse) the feature to search the path for a cloud sdk. This was because the default behavior was to download an sdk when none was specified.

On that note, I have previously seen the cloud sdk update mechanism fail because of a corrupt install directory... How it got corrupted I do not know. Perhaps that is something we should detect?  https://github.com/GoogleCloudPlatform/app-gradle-plugin/issues/334
